File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.dart
class Solution {
    List<String> letter = [
        "", "", "abc", "def", "ghi", 
        "jkl", "mno", "pqrs", "tuv", "wxyz"
    ];
  List<String> letterCombinations(String digits) {
    List<String> res = [];
    for(var i=0;i<letter[int.parse(digits[0])].length;i++){
        res.add(letter[int.parse(digits[0])][i]);
    }
    for(var i=1;i<digits.length;i++){
        final size = res.length;
        List<String> temp = [];
        for(var j=0;j<size;j++){
            for(var k=0;k<letter[int.parse(digits[i])].length;k++){
                temp.add(res[j] + letter[int.parse(digits[i])][k]);
            }
        }
        res = temp;
    }
    return res;
  }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        queue<string> q;
        string phone[] = {
            "", "", "abc", "def",
            "ghi", "jkl", "mno",
            "pqrs", "tuv", "wxyz"
        };
        for(char c : digits){
            int digit = c - '0';
            int size = q.size();
            if(!size){
                for(char letter : phone[digit]){
                    string str(1, letter);
                    q.push(str);
                }
            }else {
                for(int i=0;i<size;i++){
                    string cur = q.front();
                    q.pop();
                    for(int j=0;j<phone[digit].size();j++){
                        q.push(cur + phone[digit][j]);
                    }
                }
            }
        }
        while(!q.empty()){
            res.push_back(q.front());
            q.pop();
        }
        return res;
    }
};